File: ros2_ws/src/camera_ros/scripts/dock_pose_publisher.py
#!/usr/bin/env python3
"""
Bridge: looks up AprilTag TF and publishes as PoseStamped
on /detected_dock_pose for opennav_docking.
Caches last known pose so docking continues when tag leaves FOV.
"""
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped
import tf2_ros


class DockPosePublisher(Node):
    def __init__(self):
        super().__init__('dock_pose_publisher')
        self.declare_parameter('dock_tag_frame', 'tag25h9:0')
        self.declare_parameter('base_frame', 'odom')

        self.tag_frame = self.get_parameter('dock_tag_frame').value
        self.base_frame = self.get_parameter('base_frame').value

        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self)
        self.pub = self.create_publisher(PoseStamped, '/detected_dock_pose', 10)
        self.timer = self.create_timer(0.1, self._tick)
        self.last_pose = None

        self.get_logger().info(
            f'Dock pose publisher: {self.base_frame} -> {self.tag_frame}')

    def _tick(self):
        now = self.get_clock().now().to_msg()
        try:
            t = self.tf_buffer.lookup_transform(
                self.base_frame, self.tag_frame, rclpy.time.Time())
            msg = PoseStamped()
            msg.header.frame_id = t.header.frame_id
            msg.header.stamp = now  # Always use current time
            msg.pose.position.x = t.transform.translation.x
            msg.pose.position.y = t.transform.translation.y
            msg.pose.position.z = t.transform.translation.z
            msg.pose.orientation = t.transform.rotation
            self.last_pose = msg
        except (tf2_ros.LookupException,
                tf2_ros.ConnectivityException,
                tf2_ros.ExtrapolationException):
            pass

        # Always publish last known pose with current timestamp
        if self.last_pose is not None:
            self.last_pose.header.stamp = now
            self.pub.publish(self.last_pose)


def main(args=None):
    rclpy.init(args=args)
    rclpy.spin(DockPosePublisher())
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_ws/src/camera_ros/src/dock_pose_publisher.cpp
/**
 * @file dock_pose_publisher.cpp
 * @brief C++ replacement for dock_pose_publisher.py
 *
 * Looks up the AprilTag TF transform and publishes it as a PoseStamped
 * on /detected_dock_pose for the opennav_docking server.
 *
 * Parameters:
 *   dock_tag_frame (string, default: "tag25h9:0") - AprilTag TF frame
 *   base_frame     (string, default: "odom")       - Reference frame
 *   publish_rate   (double, default: 10.0)         - Hz
 */

#include <chrono>
#include <memory>
#include <string>

#include "geometry_msgs/msg/pose_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"

using namespace std::chrono_literals;

class DockPosePublisher : public rclcpp::Node
{
public:
  DockPosePublisher()
  : Node("dock_pose_publisher")
  {
    this->declare_parameter("dock_tag_frame", "tag25h9:0");
    this->declare_parameter("base_frame", "odom");
    this->declare_parameter("publish_rate", 10.0);

    tag_frame_   = this->get_parameter("dock_tag_frame").as_string();
    base_frame_  = this->get_parameter("base_frame").as_string();
    double rate  = this->get_parameter("publish_rate").as_double();

    tf_buffer_   = std::make_unique<tf2_ros::Buffer>(this->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>(
      "/detected_dock_pose", 10);

    auto period = std::chrono::milliseconds(static_cast<int>(1000.0 / rate));
    timer_ = this->create_wall_timer(
      period, std::bind(&DockPosePublisher::tick, this));

    RCLCPP_INFO(this->get_logger(),
      "Dock pose publisher started: %s -> %s @ %.1f Hz",
      base_frame_.c_str(), tag_frame_.c_str(), rate);
  }

private:
  void tick()
  {
    try {
      auto tf = tf_buffer_->lookupTransform(
        base_frame_, tag_frame_, tf2::TimePointZero);

      // Update cached pose on successful detection
      last_pose_.header.frame_id  = base_frame_;
      last_pose_.pose.position.x  = tf.transform.translation.x;
      last_pose_.pose.position.y  = tf.transform.translation.y;
      last_pose_.pose.position.z  = tf.transform.translation.z;
      last_pose_.pose.orientation = tf.transform.rotation;
      has_pose_ = true;
    } catch (const tf2::TransformException & ex) {
      // Tag left FOV during close approach — keep publishing last known pose
      RCLCPP_DEBUG(this->get_logger(), "TF lookup failed: %s", ex.what());
    }

    // Always publish last known pose with current timestamp (mirrors Python behaviour)
    if (has_pose_) {
      last_pose_.header.stamp = this->get_clock()->now();
      pub_->publish(last_pose_);
    }
  }

  std::string tag_frame_;
  std::string base_frame_;
  bool        has_pose_{false};

  geometry_msgs::msg::PoseStamped                          last_pose_;
  std::unique_ptr<tf2_ros::Buffer>                         tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener>              tf_listener_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr                             timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<DockPosePublisher>());
  rclcpp::shutdown();
  return 0;
}
